src/main/java/com/ldtteam/structurize/event/ClientEventSubscriber.java

```diff
@@ -151,15 +166,87 @@ public static void onClientTickEvent(final ClientTickEvent event)
             return;
         }
 
-        Minecraft.getInstance().getProfiler().push("structurize");
+        final Minecraft mc = Minecraft.getInstance();
+        mc.getProfiler().push("structurize");
 
-        if (Minecraft.getInstance().level != null && Minecraft.getInstance().level.getGameTime() % (Constants.TICKS_SECOND * 5) == 0)
+        if (mc.level != null && mc.level.getGameTime() % (Constants.TICKS_SECOND * 5) == 0)
         {
-            Minecraft.getInstance().getProfiler().push("blueprint_manager_tick");
+            mc.getProfiler().push("blueprint_manager_tick");
             BlueprintHandler.getInstance().cleanCache();
-            Minecraft.getInstance().getProfiler().pop();
+            mc.getProfiler().pop();
+        }
+
+        if (ModKeyMappings.TELEPORT.consumeClick() && mc.level != null && mc.player != null &&
+            mc.player.getMainHandItem().getItem() instanceof ItemScanTool tool)
+        {
+            if (tool.onTeleport(mc.player, mc.player.getMainHandItem()))
+            {
+                Network.getNetwork().sendToServer(new ScanToolTeleportMessage());
+            }
+        }
+
+        mc.getProfiler().pop();
+    }
+
+    @SubscribeEvent
+    public static void onMouseButtonClick(final InputEvent.MouseButton.Pre event)
+    {
+        // it would have been nice to use the "pick block" event instead, but that only fires if clicked on a block
```

Comment:
well, since u speaking about u might use pick block keybinding value instead of glfw_mouse_button_3

Reply:
I considered that, but I worried it might be set to a non-mouse-button, and then things get a lot more complicated.

Reply:
Well, do that in client tick then :D generally speaking I thought that input events are for modifying/stopping inputs before any logic, thus u should not? use them for logic

Reply:
The logic it does is to decide whether to let it continue to do the regular pick block or not, so it's mostly necessary there.  The actual work is passed off to the server via packet though.

Reply:
I've changed it now to use the block pick keybind, and `Screen.hasControlDown()`.